table of contents

Sonnen2blume opened the pull request:
I implemented a new feature: a table of contents.

A pdf is generated next to the yaml file. It contains a table with card number, title of the card or first file of the card, number of files and audio length for this card. It's a basic version to start with.

Limitations:
* there is one error left from mypy. I have no clue how to solve it
* I'm not sure if I got everything right in poetry
* there is no possibility for configuration
* tests are not implemented yet

What do you think?

mxmehl commented on tonuino_cards_manager/main.py:
```diff
@@ -78,7 +81,12 @@ def main():
         )
 
-        qrdata.append(f"{card_bytecode};{card.create_carddesc(cardno)}")
+        card_description = card.create_carddesc(cardno)
+        qrdata.append(f"{card_bytecode};{card_description}")
+        card_description = card_description.split("(")
+        card_description = card_description[1].split(")")
+        audiolength_str = str(timedelta(seconds=sum(audiolength)))
+        mylist.append([cardno, card_description[0], len(audiolength), audiolength_str])
```
Comments, and suggestions for more clear variable names

```suggestion
        card_description = card.create_carddesc(cardno)
        # Add card to QR code generation
        qrdata.append(f"{card_bytecode};{card_description}")
        # Extract content of card
        card_description = card_description.split("(")
        card_description = card_description[1].split(")")
        card_total_audio_length = str(timedelta(seconds=sum(card_audiolengths)))
        mylist.append([cardno, card_description[0], len(card_audiolengths), card_total_audio_length])
```

mxmehl commented on tonuino_cards_manager/main.py:
```diff
@@ -66,7 +69,7 @@ def main():
 
         # Create dir for card, parse sources, and copy accordingly
-        card.process_card(args.destination, config.sourcebasedir, config.filenametype)
+        audiolength = card.process_card(args.destination, config.sourcebasedir, config.filenametype)
```
```suggestion
        card_audiolengths = card.process_card(args.destination, config.sourcebasedir, config.filenametype)
```

mxmehl commented on tonuino_cards_manager/_card.py:
```diff
@@ -116,7 +117,7 @@ def check_too_many_files(self):
 
-    def process_card(self, destination: str, sourcebasepath: str, filenametype: str) -> None:
+    def process_card(self, destination: str, sourcebasepath: str, filenametype: str) -> list[int]:
         """Process a card with its configuration, also copying files. Return processed sources"""
```
```suggestion
        """Process a card with its configuration, also copying files. Return list of audio lengths of card"""
```

mxmehl commented:
Please note that I will be offline for ~1 month starting tomorrow, so I cannot review any changes before my return, sorry.

Sonnen2blume commented:
Thank you for your detailed review. I will look into it In the mean time

Sonnen2blume commented on tonuino_cards_manager/_helpers.py:
```diff
@@ -69,6 +75,12 @@ def decimal_to_hex(number: int | str) -> str:
     return f"{int(number):02x}"
 
 
+def get_audio_length(mp3file: Path) -> int:
+    """Get the audiolength of an audiofile"""
+    file = MP3(mp3file)
+    return round(file.info.length)
```
Done 🙂

Sonnen2blume replied:
Done 🙂

Sonnen2blume commented:
I implemented a basic testing oft the new functions:

1. get_audio_length
I implemented a test case for the if path. I didn't find an example for the "else" path. A non audio file creates a different error.

2. table_of_contents
The test case checks if the pdf-file is created. I would like to check the contend of the pdf as well. But how?
a. The variable "elements" contains the contend. Is it reasonable to return this variable for testing only?
b. Parse the created pdf back to python. Not sure if this is worth the effort.

Which variant would you prefer or recommend? Or is there an other option?
